File: RuntimeSources/Node/bootstrap.js
"use strict";

const fs = require("fs");
const path = require("path");
const Module = require("module");

const scriptPath = process.env.OLLAMAKIT_SCRIPT_PATH;
const inputPath = process.env.OLLAMAKIT_INPUT_PATH;
const resultPath = process.env.OLLAMAKIT_RESULT_PATH;
const workspaceRoot = process.env.OLLAMAKIT_WORKSPACE_ROOT || process.cwd();
const allowlistPath = process.env.OLLAMAKIT_NODE_ALLOWLIST_PATH || "";

function loadAllowlist() {
  try {
    if (!allowlistPath) {
      return { allowedAddons: [] };
    }
    return JSON.parse(fs.readFileSync(allowlistPath, "utf8"));
  } catch {
    return { allowedAddons: [] };
  }
}

function blockRestrictedModules(allowlist) {
  const blocked = new Set(["child_process", "cluster", "worker_threads"]);
  const originalLoad = Module._load;
  Module._load = function patchedLoad(request, parent, isMain) {
    if (blocked.has(request)) {
      throw new Error(`Module '${request}' is unavailable inside the embedded Node runtime.`);
    }
    if (request.endsWith(".node")) {
      if (!allowlist.allowedAddons || !allowlist.allowedAddons.includes(path.basename(request))) {
        throw new Error(`Native addon '${request}' is not allowlisted in this build.`);
      }
    }
    return originalLoad.call(this, request, parent, isMain);
  };
  Module._extensions[".node"] = function nativeAddonLoader(_module, filename) {
    if (!allowlist.allowedAddons || !allowlist.allowedAddons.includes(path.basename(filename))) {
      throw new Error(`Native addon '${filename}' is not allowlisted in this build.`);
    }
    throw new Error("Allowlisted native addons are not wired yet in the embedded runtime.");
  };
}

async function main() {
  const stdoutLines = [];
  const stderrLines = [];
  const originalLog = console.log;
  const originalError = console.error;

  console.log = (...args) => stdoutLines.push(args.map((value) => String(value)).join(" "));
  console.error = (...args) => stderrLines.push(args.map((value) => String(value)).join(" "));

  let payload = {
    success: false,
    stdout: "",
    stderr: "",
    exitCode: 1,
    durationMs: 0,
    result: null,
    artifacts: [],
    error: null,
  };

  try {
    const allowlist = loadAllowlist();
    blockRestrictedModules(allowlist);
    process.chdir(workspaceRoot);

    const inputValue = inputPath ? JSON.parse(fs.readFileSync(inputPath, "utf8")) : null;
    const source = fs.readFileSync(scriptPath, "utf8");
    const syntheticEntry = path.join(workspaceRoot, "__ollamakit__.js");
    const localRequire = Module.createRequire(syntheticEntry);
    const dirname = path.dirname(scriptPath);
    const moduleShim = { exports: {} };
    const executionContext = {
      artifacts: [],
      result: undefined,
    };
    const originalResult = globalThis.result;
    const originalArtifacts = globalThis.artifacts;
    globalThis.result = undefined;
    globalThis.artifacts = executionContext.artifacts;
    const runner = new Function(
      "input",
      "workspaceRoot",
      "require",
      "module",
      "exports",
      "__filename",
      "__dirname",
      "context",
      `${source}
return typeof globalThis.result === "undefined" ? module.exports : globalThis.result;`
    );
    try {
      const result = await runner(
        inputValue,
        workspaceRoot,
        localRequire,
        moduleShim,
        moduleShim.exports,
        scriptPath,
        dirname,
        executionContext
      );

      payload.success = true;
      payload.exitCode = 0;
      payload.result = result;
      payload.artifacts = Array.isArray(globalThis.artifacts) ? globalThis.artifacts : executionContext.artifacts;
    } finally {
      if (typeof originalResult === "undefined") {
        delete globalThis.result;
      } else {
        globalThis.result = originalResult;
      }

      if (typeof originalArtifacts === "undefined") {
        delete globalThis.artifacts;
      } else {
        globalThis.artifacts = originalArtifacts;
      }
    }
  } catch (error) {
    stderrLines.push(error && error.stack ? error.stack : String(error));
    payload.error = error ? String(error.message || error) : "Unknown embedded Node failure.";
  } finally {
    console.log = originalLog;
    console.error = originalError;
  }

  payload.stdout = stdoutLines.join("\n");
  payload.stderr = stderrLines.join("\n");
  fs.writeFileSync(resultPath, JSON.stringify(payload));
}

main().then(
  () => process.exit(0),
  (error) => {
    const payload = {
      success: false,
      stdout: "",
      stderr: error && error.stack ? error.stack : String(error),
      exitCode: 1,
      durationMs: 0,
      result: null,
      artifacts: [],
      error: error ? String(error.message || error) : "Unknown embedded Node failure.",
    };
    fs.writeFileSync(resultPath, JSON.stringify(payload));
    process.exit(1);
  }
);

File: RuntimeSources/Python/bootstrap.py
import contextlib
import importlib.machinery
import io
import json
import os
import sys
import traceback
from pathlib import Path


def load_allowlist(path: str) -> dict:
    try:
        with open(path, "r", encoding="utf-8") as handle:
            return json.load(handle)
    except Exception:
        return {"allowed_modules": []}


def main() -> int:
    script_path = Path(os.environ["OLLAMAKIT_SCRIPT_PATH"])
    input_path = Path(os.environ["OLLAMAKIT_INPUT_PATH"])
    result_path = Path(os.environ["OLLAMAKIT_RESULT_PATH"])
    workspace_root = Path(os.environ.get("OLLAMAKIT_WORKSPACE_ROOT", "")).resolve()
    allowlist_path = os.environ.get("OLLAMAKIT_PYTHON_ALLOWLIST_PATH", "")

    payload = {
        "success": False,
        "stdout": "",
        "stderr": "",
        "exitCode": 1,
        "durationMs": 0,
        "result": None,
        "artifacts": [],
        "error": None,
    }

    input_value = None
    if input_path.exists():
        try:
            input_value = json.loads(input_path.read_text(encoding="utf-8"))
        except Exception as exc:
            payload["error"] = f"Failed to decode input JSON: {exc}"
            result_path.write_text(json.dumps(payload), encoding="utf-8")
            return 0

    allowlist = load_allowlist(allowlist_path) if allowlist_path else {"allowed_modules": []}
    if not allowlist.get("allowed_modules"):
        try:
            importlib.machinery.EXTENSION_SUFFIXES[:] = []
        except Exception:
            pass

    if workspace_root and workspace_root.exists():
        os.chdir(workspace_root)
        sys.path.insert(0, str(workspace_root))
        for candidate in [
            workspace_root / "site-packages",
            workspace_root / ".venv" / "lib",
            *sorted((workspace_root / ".venv" / "lib").glob("python*/site-packages")),
        ]:
            if candidate.exists():
                sys.path.insert(0, str(candidate))

    stdout_capture = io.StringIO()
    stderr_capture = io.StringIO()
    globals_dict = {
        "__name__": "__main__",
        "__file__": str(script_path),
        "input": input_value,
        "workspace_root": str(workspace_root),
        "result": None,
        "artifacts": [],
    }

    try:
        code = script_path.read_text(encoding="utf-8")
        with contextlib.redirect_stdout(stdout_capture), contextlib.redirect_stderr(stderr_capture):
            exec(compile(code, str(script_path), "exec"), globals_dict, globals_dict)
        payload["success"] = True
        payload["exitCode"] = 0
        payload["result"] = globals_dict.get("result")
        payload["artifacts"] = globals_dict.get("artifacts") or []
    except BaseException as exc:  # noqa: BLE001
        stderr_capture.write(traceback.format_exc())
        payload["error"] = str(exc)
        payload["exitCode"] = 1

    payload["stdout"] = stdout_capture.getvalue()
    payload["stderr"] = stderr_capture.getvalue()
    result_path.write_text(json.dumps(payload), encoding="utf-8")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
